tests/hash/main.cpp

```diff
@@ -127,3 +161,134 @@ int main(int argc, char *argv[]) {
   }
   return error ? 1 : 0;
 }
+
+#if defined(__x86_64__) || (defined(_MSC_VER) && defined(_WIN64))
+
+#include <emmintrin.h>
+
+#if defined(_MSC_VER) || defined(__MINGW32__)
+  #include <intrin.h>
+#else
+  #include <wmmintrin.h>
+#endif
+
+#endif
+
+static inline bool test_variant2_int_sqrt_sse(const uint64_t sqrt_input, const uint64_t correct_result)
+{
+#if defined(__x86_64__) || (defined(_MSC_VER) && defined(_WIN64))
+  uint64_t sqrt_result;
+  VARIANT2_INTEGER_MATH_SQRT_STEP_SSE2();
+  VARIANT2_INTEGER_MATH_SQRT_FIXUP(sqrt_result);
+  if (sqrt_result != correct_result) {
+    cerr << "Integer sqrt (SSE2 version) returned incorrect result for N = " << sqrt_input << endl;
+    cerr << "Expected result: " << correct_result << endl;
+    cerr << "Returned result: " << sqrt_result << endl;
+    return false;
+  }
+#endif
+
+  return true;
+}
+
+static inline bool test_variant2_int_sqrt_fp64(const uint64_t sqrt_input, const uint64_t correct_result)
+{
+#if defined DBL_MANT_DIG && (DBL_MANT_DIG >= 50)
+  uint64_t sqrt_result;
+  VARIANT2_INTEGER_MATH_SQRT_STEP_FP64();
+  VARIANT2_INTEGER_MATH_SQRT_FIXUP(sqrt_result);
+  if (sqrt_result != correct_result) {
+    cerr << "Integer sqrt (FP64 version) returned incorrect result for N = " << sqrt_input << endl;
+    cerr << "Expected result: " << correct_result << endl;
+    cerr << "Returned result: " << sqrt_result << endl;
+    return false;
+  }
+#endif
+
+  return true;
+}
+
+static inline bool test_variant2_int_sqrt_ref(const uint64_t sqrt_input, const uint64_t correct_result)
+{
+  uint64_t sqrt_result;
+  VARIANT2_INTEGER_MATH_SQRT_STEP_REF();
+  if (sqrt_result != correct_result) {
+    cerr << "Integer sqrt (reference version) returned incorrect result for N = " << sqrt_input << endl;
+    cerr << "Expected result: " << correct_result << endl;
+    cerr << "Returned result: " << sqrt_result << endl;
+    return false;
+  }
+
+  return true;
+}
+
+static inline bool test_variant2_int_sqrt(const uint64_t sqrt_input, const uint64_t correct_result)
+{
+  if (!test_variant2_int_sqrt_sse(sqrt_input, correct_result)) {
+    return false;
+  }
+  if (!test_variant2_int_sqrt_fp64(sqrt_input, correct_result)) {
+    return false;
+  }
+
+  // Reference version is integer only and doesn't have any undefined behavior
+  // We don't need to test it every time
+  // Uncomment this if you're modifying/optimizing the reference version
+  //if (!test_variant2_int_sqrt_ref(sqrt_input, correct_result)) {
```

Comment:
Nobody will ever see that there is a test for the integer sqrt version available even if there is a comment in the code. This test shoud be enabled by default and the reference that this test is disabled should be removed from the code. 

The needed time for this test should be small enough to have this test enabled by default.

Reply:
It's much slower than floating point versions. There are 7116134816 edge cases to test - it won't be possible to test them all. I will add a code that runs reference implementation for 1% of them, including 0, 2^64-1 and other extreme cases. It should still catch all implementation bugs.